### PR TITLE
Add updaters for Amazon Linux 2018.03 and Amazon Linux 2

### PR DESCRIPTION
**Pull request by ericysim**
We get vulnerabilities from ALAS (Amazon Linux Security Advisories) data, which can be found in updateinfo.xml from the repos.

---

Here's ALAS data in a human-readable format: https://alas.aws.amazon.com/

We don't use https://alas.aws.amazon.com/ for the updater. Instead, we use updateinfo.xml from the repos, which has the same information (and more) in XML format.

Here's how to get the URL of updateinfo.xml.gz. We have to do this because the hash in the URL changes whenever there's a push.

> - Get the mirror.list URL, which is found in the repo definitions in /etc/yum.repos.d. (In the code, this URL is hardcoded.)
>   - e.g. http://repo.us-west-2.amazonaws.com/2018.03/updates/x86_64/mirror.list
> - Get the first line of mirror.list; this is the base URL.
>   - e.g. http://packages.us-west-2.amazonaws.com/2018.03/updates/9f831135fa79/x86_64
> - Append repodata/repomd.xml to the base URL. This XML file defines the metadata available in the repo.
>   - e.g. http://packages.us-west-2.amazonaws.com/2018.03/updates/9f831135fa79/x86_64/repodata/repomd.xml
> - Find a data element with type="updateinfo" and follow the location relative to the base URL.
>    - e.g. http://packages.us-west-2.amazonaws.com/2018.03/updates/9f831135fa79/x86_64/repodata/updateinfo.xml.gz

We do this process in the code.

Question: I'd like the mirror.list URLs to be user-configurable (e.g. http://repo.us-east-2.amazonaws.com/2018.03/updates/x86_64/mirror.list instead of http://repo.us-west-2.amazonaws.com/2018.03/updates/x86_64/mirror.list). What's a good way to do this?

### CODE REVIEW

**Comment by KeyboardNerd** [comment]
Currently, the updaters are not configurable. We can make updaters configurable per updater base in the future, but it's not the current goal.

**Comment by ericysim** [comment]
I see; it's no problem :)

File: ext/vulnsrc/amzn/amzn.go [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+	"bufio"
+	"compress/gzip"
+	"encoding/xml"
+	"io"
+	"regexp"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"fmt"
```
Comment (jzelinskie) [inline ext/vulnsrc/amzn/amzn.go]: this should be with all the other stdlib imports

File: ext/vulnsrc/amzn/amzn.go [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+	"github.com/coreos/clair/pkg/httputil"
+)
+
+const (
+	amazonLinux1UpdaterFlag   = "amazonLinux1Updater"
+	amazonLinux1MirrorListURI = "http://repo.us-west-2.amazonaws.com/2018.03/updates/x86_64/mirror.list"
+	amazonLinux1Name          = "Amazon Linux 2018.03"
+	amazonLinux1Namespace     = "amzn:2018.03"
+	amazonLinux1LinkFormat    = "https://alas.aws.amazon.com/%s.html"
+	amazonLinux2UpdaterFlag   = "amazonLinux2Updater"
```
Comment (jzelinskie) [inline ext/vulnsrc/amzn/amzn.go:45]: put an extra new line above this line so that AL1 has its own block and AL2 has its own block

**Comment by subash-canapathy** [comment]
Can someone tag an approver to review this. We want to be enable scans for amazon linux containers

**Review by subash-canapathy — APPROVED** [review]
Looks good to me, passes tests.

File: ext/vulnsrc/amzn/amzn.go [hunk 3]
```diff
@@ -0,0 +1,348 @@
+// Copyright 2017 clair authors
```
Comment (jzelinskie) [inline ext/vulnsrc/amzn/amzn.go]: It's not 2017 anymore!

**Comment by ericysim** [comment]
How can we get traction on this pull request? @jzelinskie @KeyboardNerd

**Comment by KeyboardNerd** [comment]
This looks really cool.
What the license is the vuln source is under? 
We'd like to enable people to use Clair in both commercial or non-commercial settings.

**Comment by iliana** [comment]
Hi, I work on Amazon Linux.

> What the license is the vuln source is under?

The vulnerability data (updateinfo.xml files in our repositories) are free to use.

If you need a more specific answer I'll need to do a query to our legal folks, but there is certainly not a non-commercial restriction on the data.

**Comment by KeyboardNerd** [comment]
Thanks, we'll review the code.

**Review by jzelinskie — CHANGES_REQUESTED** [review]
I'm really sorry for the delay on the review. Honestly, I only have very small style nitpicks. The rest looks great! There's a lot of rpm-based updaters, we should eventually write some shared plumbing for them, but for right now this is perfect.

edit: actually one more thing, you need to add Amazon Linux in the documentation for supported platforms along with the licensed for the data being used.

File: ext/vulnsrc/amzn/amzn.go (hunk 1 above)
Comment (ericysim) [reply ext/vulnsrc/amzn/amzn.go]: done

File: ext/vulnsrc/amzn/amzn.go (hunk 2 above)
Comment (ericysim) [reply ext/vulnsrc/amzn/amzn.go:45]: done

File: ext/vulnsrc/amzn/amzn.go (hunk 3 above)
Comment (ericysim) [reply ext/vulnsrc/amzn/amzn.go]: done

**Comment by ericysim** [comment]
Thanks @jzelinskie, I made the recommended changes.

**Comment by ericysim** [comment]
> edit: actually one more thing, you need to add Amazon Linux in the documentation for supported platforms along with the licensed for the data being used.

Ok, will do. I will ask @iliana about the license again.

**Comment by subash-canapathy** [comment]
@ericysim - Hi Eric is this now ready to be merged.

**Comment by ericysim** [comment]
Hi @subash-canapathy, I got the license info this morning. I'm going to add Amazon Linux to the docs and push a commit by the end of the day :)

**Comment by ericysim** [comment]
Hi @jzelinskie, I made the recommended changes.

**Review by jzelinskie — APPROVED** [review]

**Pull request merged**